test/integration/protocol_integration_test.cc

```diff
@@ -747,6 +761,251 @@ TEST_P(DownstreamProtocolIntegrationTest, LargeRequestHeadersAccepted) {
+// once after iteration is resumed.
+TEST_P(DownstreamProtocolIntegrationTest, testDecodeHeadersReturnsStopAll) {
+  config_helper_.addFilter(R"EOF(
+name: call-decodedata-once-filter
+)EOF");
+  config_helper_.addFilter(R"EOF(
+name: decode-headers-return-stop-all-filter
+)EOF");
+  config_helper_.addFilter(R"EOF(
+name: passthrough-filter
+)EOF");
+
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Sends a request with headers and data.
+  changeHeadersForStopAllTests(default_request_headers_, false);
+  auto encoder_decoder = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+  for (int i = 0; i < count_ - 1; i++) {
+    codec_client_->sendData(*request_encoder_, size_, false);
+  }
+  // Sleeps for 1s in order to be consistent with testDecodeHeadersReturnsStopAllWatermark.
+  sleep(1);
+  codec_client_->sendData(*request_encoder_, size_, true);
+  waitForNextUpstreamRequest();
+
+  upstream_request_->encodeHeaders(default_response_headers_, true);
+  response->waitForEndStream();
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ(count_ * size_ + added_decoded_data_size_, upstream_request_->bodyLength());
+  EXPECT_EQ(true, upstream_request_->complete());
+
+  // Sends a request with headers, data, and trailers.
+  auto encoder_decoder_2 = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder_2.first;
+  response = std::move(encoder_decoder_2.second);
+  for (int i = 0; i < count_; i++) {
+    codec_client_->sendData(*request_encoder_, size_, false);
+  }
+  Http::TestHeaderMapImpl request_trailers{{"trailer", "trailer"}};
+  codec_client_->sendTrailers(*request_encoder_, request_trailers);
+  waitForNextUpstreamRequest();
+
+  upstream_request_->encodeHeaders(default_response_headers_, true);
+  response->waitForEndStream();
+  if (downstreamProtocol() == Http::CodecClient::Type::HTTP2) {
+    // decode-headers-return-stop-all-filter calls addDecodedData in decodeData and decodeTrailers.
+    // 2 decoded data were added.
+    EXPECT_EQ(count_ * size_ + added_decoded_data_size_ * 2, upstream_request_->bodyLength());
+  } else {
+    EXPECT_EQ(count_ * size_ + added_decoded_data_size_ * 1, upstream_request_->bodyLength());
+  }
+  EXPECT_EQ(true, upstream_request_->complete());
+}
+
+// Tests StopAllIterationAndWatermark. decode-headers-return-stop-all-watermark-filter sets buffer
+// limit to 100. Verifies data pause when limit is reached, and resume after iteration continues.
+TEST_P(DownstreamProtocolIntegrationTest, testDecodeHeadersReturnsStopAllWatermark) {
+  config_helper_.addFilter(R"EOF(
+name: decode-headers-return-stop-all-filter
+)EOF");
+  config_helper_.addFilter(R"EOF(
+name: passthrough-filter
+)EOF");
+
+  // Sets initial stream window to min value to make the client sensitive to a low watermark.
```

Comment:
s/to make the client sensitive to a low watermark/to make sure the client initially send less data than the buffer_limit.

Reply:
I think it may not guarantee the client sends less than buffer_limit. I think when watermark is reached, we stop reading, but client can still send, :)

Reply:
When buffer_limit is reached, we stop reading. But it can't guarantee the client sends less than buffer_limit.